chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala

```diff
@@ -500,87 +503,161 @@ case class ChainHandler(
-        for {
-          headersToCalc <-
-            blockHeaderDAO.getBetweenHeights(batchStartHeight, batchEndHeight)
-          sortedHeaders = headersToCalc.sortBy(_.height)
-          headersWithWork <- loop(sortedHeaders, prev)
-          next <- loop2(maxHeight, headersWithWork)
-        } yield next
+    Future
+      .sequence(batchesNested)
+      .map(_.flatten)
+  }
+
+  private def runRecalculateChainWork(
+      maxHeight: Int,
+      lastHeader: BlockHeaderDb): Future[Vector[BlockHeaderDb]] = {
+    val currentHeight = lastHeader.height
+    val numBatches = 5
+    val batchSize =
+      chainConfig.appConfig.chain.difficultyChangeInterval / numBatches
+    if (currentHeight >= maxHeight) {
+      Future.successful(Vector.empty)
+    } else {
+      val batchStartHeight = currentHeight + 1
+
+      val headersToCalcF = batchAndGetHeaders(
+        batchSize = batchSize,
+        batchStartHeight = batchStartHeight,
+        maxHeight = maxHeight,
+        numBatches = numBatches
+      )
+
+      for {
+        headersToCalc <- headersToCalcF
+        sortedHeaders = headersToCalc.sortBy(_.height)
+        _ = sortedHeaders.headOption.map { h =>
+          logger.info(
+            s"Recalculating chain work... current height: ${h.height}")
+        }
+        headersWithWork = calcChainWork(remainingHeaders = sortedHeaders,
+                                        accum = Vector.empty,
+                                        lastHeaderWithWorkInDb = lastHeader)
+
+        //unfortunately on sqlite there is a bottle neck here
+        //sqlite allows you to read in parallel but only write
+        //sequentially https://stackoverflow.com/a/23350768/967713
+        //so while it looks like we are executing in parallel
+        //in reality there is only one thread that can write to the db
+        //at a single time
+        _ = logger.debug(
+          s"Upserting from height=${headersWithWork.head.height} to height=${headersWithWork.last.height}")
+        _ <- FutureUtil.batchExecute(
+          headersWithWork,
+          blockHeaderDAO.upsertAll,
+          Vector.empty,
+          batchSize
+        )
+        _ = logger.debug(
+          s"Done upserting from height=${headersWithWork.head.height} to height=${headersWithWork.last.height}")
+        next <- runRecalculateChainWork(maxHeight, headersWithWork.last)
+      } yield {
+        next
       }
     }
+  }
 
-    for {
-      maxHeight <- blockHeaderDAO.maxHeight
-      startHeight <- blockHeaderDAO.getLowestNoWorkHeight
-      start <-
+  def recalculateChainWork: Future[ChainHandler] = {
+    logger.info("Calculating chain work for previous blocks")
+
+    val maxHeightF = blockHeaderDAO.maxHeight
+    val startHeightF = blockHeaderDAO.getLowestNoWorkHeight
+    val startF = for {
+      startHeight <- startHeightF
+      headers <- {
         if (startHeight == 0) {
-          Future.successful(Vector.empty)
```

Comment:
We want to be able to recalc on the gensis header, if you have an old db it will be `0`